File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        nodeprev = []
        def recurse(node, level):
            if node == None:
                return
            if len(nodeprev) == level:
                nodeprev.append(Node())
            nodeprev[level].next = node
            nodeprev[level] = nodeprev[level].next
            recurse(node.left, level+1)
            recurse(node.right, level+1)
        recurse(root, 0)
        return root
        

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        queue<Node*> nodeque;
        if (root != nullptr)
            nodeque.push(root);
        while (!nodeque.empty()) {
            int size = nodeque.size();
            Node* prev = new Node();
            for (int i = 0; i < size; i++) {
                Node* cur = nodeque.front();
                nodeque.pop();
                prev->next = cur;
                prev = prev->next;
                if (cur->left)
                    nodeque.push(cur->left);
                if (cur->right)
                    nodeque.push(cur->right);
            }
            prev->next = nullptr;
        }
        return root;
    }
};